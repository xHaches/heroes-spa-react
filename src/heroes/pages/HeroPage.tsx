import { useMemo } from "react"
import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../helpers"

export const HeroPage = () => {
  
  const { customIdPersonalizadoDelHeroe } = useParams()
  
  // useMemo -> memorizar valores useCallback -> memorizar funciones
  const hero = useMemo(() => getHeroById(customIdPersonalizadoDelHeroe!), [customIdPersonalizadoDelHeroe]) ;
  
  const navigate = useNavigate();
  const onBack = () => {
      navigate(-1)
  }

  if(!hero) {
    return <Navigate to="/marvel" />
  }

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img 
          src={`/assets/heroes/${customIdPersonalizadoDelHeroe}.jpg`}
          alt={hero.superhero}
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>
      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item"><b>Alter ego:</b> {hero.alter_ego}</li>
          <li className="list-group-item"><b>Publisher:</b> {hero.publisher}</li>
          <li className="list-group-item"><b>First appeareance:</b> {hero.first_appearance}</li>
        </ul>
        <h5 className="mt-3"> Characters </h5>
        <p>{hero.characters}</p>
        <button className="btn btn-outline-primary"
          onClick={onBack}
        >
          Back
        </button>
      </div>
    </div>
  )
}
