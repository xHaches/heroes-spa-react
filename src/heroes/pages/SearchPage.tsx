import { FormEvent } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import queryString from 'query-string';
import { HeroCard } from "../components/HeroCard"
import { getHeroesByName } from "../helpers";

export const SearchPage = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const {q = '' } = queryString.parse(location.search);

  const heroes = getHeroesByName(q as string);

  const showSearch = (q?.length === 0);
  const shorError = (q?.length! > 0) && (heroes.length === 0);

  const { searchText, onInputChange } = useForm<{searchText: string}>({
    searchText: q as string
  });

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    navigate(`?q=${searchText}`);
  }

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={handleSubmit} aria-label="form">
            <input 
              type="text" 
              placeholder="Search a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
            <button className="btn btn-outline-primary mt-1">
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {/* {
            (q === '')
            ? <div className="alert alert-primary">Search a hero</div>
            : (heroes.length === 0) 
            && <div className="alert alert-danger">No hero with <b>{q}</b></div>
          } */}
          <div className="alert alert-primary animate__animated animate__fadeInLeft" style={{display: showSearch ? "" : "none"}} aria-label="input-empty">Search a hero</div>
          <div className="alert alert-danger animate__animated animate__fadeInLeft" style={{display: shorError ? "" : "none"}} aria-label="input-not-found">No hero with <b>{q}</b></div>
          {
            heroes.map(hero => {
              return <HeroCard key={hero.id} {...hero} />
            })
          }
        </div>
      </div>
    </>
  )
}
