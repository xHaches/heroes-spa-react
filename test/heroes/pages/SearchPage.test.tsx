import React from 'react'
import { describe, test, expect, afterEach, vi } from 'vitest';
import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom'
import { SearchPage } from '../../../src/heroes/pages/SearchPage';

// se usa como referencia para un to haveBeenCalled en los tests
const mockedUseNavigate = vi.fn();

vi.mock('react-router-dom', async() => {
    const actual = await vi.importActual('react-router-dom');
    return {
        ...actual as any,
        useNavigate: () => mockedUseNavigate
    }
})

describe('Pruebas en <SearchPage />', () => {

    afterEach(() => {
        vi.clearAllMocks();
        cleanup();
    });


    test('debe de mostrarse correctamente con valores por defecto', () => {
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
    });

    test('debe de mostrar a Batman y el input con el valor del QueryString', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        );
        const input = screen.getByRole<HTMLInputElement>('textbox');
        expect(input.value).toBe('batman');
        const img = screen.getByRole<HTMLImageElement>('img');
        expect(img.src).toContain('/assets/heroes/dc-batman.jpg');
        const heroEmpty = screen.getByLabelText<HTMLDivElement>('input-empty');
        expect(heroEmpty.style.display).toBe('none')
    });

    test('debe de mostrar un error si no se encuentra el hero (batman123)', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        );
        const heroError = screen.getByLabelText<HTMLDivElement>('input-not-found');
        expect(heroError.style.display).toBe('')
    });

    test('debe de llamar el navigate a la pantalla nueva', () => {
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );
        const input = screen.getByRole('textbox');
        fireEvent.change(input, { target: {name: 'searchText', value: 'superman'}});
        const form = screen.getByLabelText('form');
        fireEvent.submit(form);
        // fireEvent.submit(input);
        expect(mockedUseNavigate).toHaveBeenCalledWith('?q=superman')
    });

});