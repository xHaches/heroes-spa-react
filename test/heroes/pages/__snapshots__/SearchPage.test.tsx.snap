// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pruebas en <SearchPage /> > debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="input-empty"
        class="alert alert-primary animate__animated animate__fadeInLeft"
      >
        Search a hero
      </div>
      <div
        aria-label="input-not-found"
        class="alert alert-danger animate__animated animate__fadeInLeft"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
